src/Config.hs

```diff
@@ -0,0 +1,223 @@
+
+import Options.Applicative (value, long, auto, option, metavar, help, flag,
+                            str, argument, many, strOption, (<**>))
+
+import qualified Options.Applicative as OptParse
+import qualified Options.Applicative.Builder.Internal as OptParse
+import qualified Text.Read as Read
+
+-- ! Type alias for enviornment variables, used for readability in this module.
+type EnvVar = (String, String)
+
+-- ! Newtype wrapper for millisecond values.
+newtype MilliSeconds = MilliSeconds { unMilliSeconds :: Int }
+  deriving (Eq, Show)
+
+-- | @Options@ contains all the configuration we support in vaultenv. It is
+-- used in our @Main@ module to specify behaviour.
+data Options = Options
+  { oVaultHost       :: String
+  , oVaultPort       :: Int
+  , oVaultToken      :: String
+  , oSecretFile      :: FilePath
+  , oCmd             :: String
+  , oArgs            :: [String]
+  , oNoConnectTls    :: Bool
+  , oNoValidateCerts :: Bool
+  , oNoInheritEnv    :: Bool
+  , oRetryBaseDelay  :: MilliSeconds
+  , oRetryAttempts   :: Int
+  } deriving (Eq, Show)
+
+-- | Behaviour flags that we allow users to set via environment variables.
+-- This type is internal to the workings of this module. It is used as an
+-- intermediate value to get optparse-applicative to play nice with environment
+-- variables as used for behavior flags.
+data EnvFlags = EnvFlags
+  { efNoConnectTls :: Bool
+  , efNoValidateCerts :: Bool
+  , efNoInheritEnv :: Bool
+  }
+
+-- | Parse program options from the command line and the process environment.
+parseOptionsFromEnvAndCli :: [EnvVar] -> IO Options
+parseOptionsFromEnvAndCli envVars =
+  let envFlags = parseEnvFlags envVars
+      parser = optionsParserWithInfo envFlags envVars
+  in OptParse.execParser parser
+
+-- | Parses behaviour flags from a list of environment variables. If an
+-- environment variable corresponding to the flag is set to @"true"@ or
+-- @"false"@, we use that as the default on the corresponding CLI option.
+--
+-- If these variables aren't present, we default to @False@. We print an error
+-- if they're set to anything else than @"true"@ or @"false"@.
+parseEnvFlags :: [EnvVar] -> EnvFlags
+parseEnvFlags envVars
+  = EnvFlags
+  { efNoConnectTls = lookupEnvFlag "VAULTENV_NO_CONNECT_TLS"
+  , efNoValidateCerts = lookupEnvFlag "VAULTENV_NO_VALIDATE_CERTS"
+  , efNoInheritEnv = lookupEnvFlag "VAULTENV_NO_INHERIT_ENV"
+  }
+  where
+    lookupEnvFlag key =
+      case lookup key envVars of
+        Just "true" -> True
+        Just "false" -> False
+        Nothing -> False
+        _ -> errorWithoutStackTrace $ "[ERROR]: Invalid value for environment variable " ++ key
+
+-- | This function adds metadata to the @Options@ parser so it can be used with
+-- execParser.
+optionsParserWithInfo :: EnvFlags -> [EnvVar] -> OptParse.ParserInfo Options
+optionsParserWithInfo envFlags localEnvVars =
+  OptParse.info
+    (optionsParser envFlags localEnvVars <**> OptParse.helper)
```

Comment:
I think `OptParse.helper <*> optionsParser envFlags localEnvVars` should probably work as well here, and  `<*>` is better known than `<**>`.

Reply:
Agreed & changed. I think this was an artifact of me following an optparse tutorial way in the beginning